carma_wm/src/CARMAWorldModel.cpp

```diff
@@ -1118,5 +1118,78 @@ std::vector<lanelet::CarmaTrafficLightPtr> CARMAWorldModel::predictTrafficLight(
+    return;
+  }
+  for (auto curr_intersection : spat_msg.intersection_state_list)
+  {
+    for (auto current_movement_state : curr_intersection.movement_list) 
+    {
+      lanelet::Id curr_light_id = getTrafficLightId(curr_intersection.id.id, current_movement_state.signal_group);
+
+      if (curr_light_id == lanelet::InvalId)
+      {
+        ROS_DEBUG_STREAM("Received a SPAT message for traffic light that is not in the map with intersection_id: " << (int)curr_intersection.id.id << 
+                           ", and signal_group_id: " << (int)current_movement_state.signal_group);
+        continue;
+      }
+      
+      auto general_regem = semantic_map_->regulatoryElementLayer.get(curr_light_id);
+      
+      auto lanelets_general = semantic_map_->laneletLayer.findUsages(general_regem);
+      if (lanelets_general.empty())
+      {
+        ROS_WARN_STREAM("Received a SPAT message for traffic light that is not owned by any lanelet with intersection_id: " << (int)curr_intersection.id.id << 
+                           ", and signal_group_id: " << (int)current_movement_state.signal_group);
+        continue;
+      }
+      auto curr_light_list = lanelets_general[0].regulatoryElementsAs<lanelet::CarmaTrafficLight>();
+      
+      if (curr_light_list.empty())
+      {
+        ROS_WARN_STREAM("There was an error querying traffic light with intersection_id: " << (int)curr_intersection.id.id << 
+                           ", and signal_group_id: " << (int)current_movement_state.signal_group);
+        continue;
+      }
+      lanelet::CarmaTrafficLightPtr curr_light = curr_light_list[0];
+      
+      // reset states if the intersection's geometry changed 
+      if (curr_light->revision_ != curr_intersection.revision)
+      {
+        ROS_INFO_STREAM("Received a new intersection geometry. intersection_id: " << (int)curr_intersection.id.id << 
+                           ", and signal_group_id: " << (int)current_movement_state.signal_group);
+        traffic_light_states_[curr_intersection.id.id].clear();
+      }
+
+      // currently expecting 1 lane, so 0th index is used only when setting states
+      if (current_movement_state.movement_event_list.empty())
+      {
+        ROS_INFO_STREAM("movement_event_list is empty . intersection_id: " << (int)curr_intersection.id.id << 
+                           ", and signal_group_id: " << (int)current_movement_state.signal_group);
+        continue;
+      }
+
+      traffic_light_states_[curr_intersection.id.id][current_movement_state.signal_group].push_back(std::make_pair(ros::Time(current_movement_state.movement_event_list[0].timing.min_end_time), 
+                              static_cast<lanelet::CarmaTrafficLightState>(current_movement_state.movement_event_list[0].event_state.movement_phase_state)));
+      
+      if (traffic_light_states_[curr_intersection.id.id][current_movement_state.signal_group].size() >= 2 
```

Comment:
I think you probably can't wait for the full phase cycle to be received before the CarmaTrafficLight is updated. It seems unlikely that the vehicle would have like 40-60s to approach the light before needing this information. I think you probably need to add in a default phase state duration which can be used until the actual time is received. That way guidance can do its best to plan with the current state until the future state is also received.